Allow importing Svg icons

alex-ds13 opened the pull request:
This PR enables the addition of custom icons to the font.

To do so we simply need to put the svg image files on the same folder as the TOML file and then on the `[glyphs]` section we set the name we want for said icon as the key and the value needs to be the file name for the svg image that will be imported for that icon.

This file name **needs** to end in ".svg" so that we recognize the entry as a custom svg icon that needs to be imported.

As an example if we wanted an `iced` icon, we simply need to add an svg image to the folder, let's say it's named `iced.svg` then on the `example-icons.toml` file we would add it like this:
```toml
module = "icon"

[glyphs]
edit = "fontawesome-pencil"
save = "entypo-floppy"
trash = "typicons-trash"
iced = "iced.svg"
```

### **Note:**
It's important that the svg file is correctly structured to be able to be imported and some svg tags will be ignored. Only the `path` will be considered, any `rect`, `circle` and others like it will simply be ignored.
Basically you can read this from fontello wiki to know more about this: https://github.com/fontello/fontello/wiki/How-to-use-custom-images#importing-svg-images

In case you are facing any issue with an svg file, first try importing it on the fontello site. If it doesn't work there either than you'll know that fontello doesn't support it and it needs to be changed like mentioned on the link above.

alex-ds13 committed:
fix: lint warning

alex-ds13 committed:
fix: lint warning

alex-ds13 committed:
refactor: move structs out of `build` and make them public

alex-ds13 committed:
refactor: move structs out of `build` and make them public

alex-ds13 committed:
feat: implement custom icons loading

- Adds a new section to the icons TOML file, named "svg_icons" where the
  user can define multiple svg images where the key will be the name of
  the icon to be created and needs to be unique and the value is the
  file name of the svg file that will be loaded (the extension can be
  omitted). This svg file needs to be on the same folder as the TOML file.

alex-ds13 committed:
feat: implement custom icons loading

- Adds a new section to the icons TOML file, named "svg_icons" where the
  user can define multiple svg images where the key will be the name of
  the icon to be created and needs to be unique and the value is the
  file name of the svg file that will be loaded (the extension can be
  omitted). This svg file needs to be on the same folder as the TOML file.

alex-ds13 committed:
refactor(svg_parser): use imports and apply lint suggestions

alex-ds13 committed:
refactor(svg_parser): use imports and apply lint suggestions

alex-ds13 committed:
fix: ignore svg percentage sizes like fontello

- We weren't processing the percentage sizes correctly before, since we
  were merely changing the size variable without actually scaling the
  path. However it doesn't make sense to scale the path accordingly to
  a percentage size only to then scale it again to 1000 as fontello
  wants it to be. So instead of scaling twice we ignore it just like
  fontello does currently.

alex-ds13 committed:
fix: ignore svg percentage sizes like fontello

- We weren't processing the percentage sizes correctly before, since we
  were merely changing the size variable without actually scaling the
  path. However it doesn't make sense to scale the path accordingly to
  a percentage size only to then scale it again to 1000 as fontello
  wants it to be. So instead of scaling twice we ignore it just like
  fontello does currently.

alex-ds13 committed:
feat: allow changes on svg files to trigger build

alex-ds13 committed:
feat: allow changes on svg files to trigger build

alex-ds13 committed:
feat: allow svg icons on `[glyphs]` section

- Svg icons can be set on the `[glyphs]` section by setting the svg icon
  file name as the value of an entry. The value needs to end in ".svg"
  for it to be considered as an svg icon import.

alex-ds13 committed:
feat: allow svg icons on `[glyphs]` section

- Svg icons can be set on the `[glyphs]` section by setting the svg icon
  file name as the value of an entry. The value needs to end in ".svg"
  for it to be considered as an svg icon import.

alex-ds13 committed:
fix: remove `svg_icons`

- Svg icons are now only set on the `[glyphs]` section.

alex-ds13 committed:
fix: remove `svg_icons`

- Svg icons are now only set on the `[glyphs]` section.

alex-ds13 commented:
We probably can remove that "wrong" `iced_logo` icon.

Also if you want I can remove the changes I've made to the `example/src/main.rs`. I've made them mostly to test that the icons were being properly imported.

alex-ds13 commented:
I've also just noticed an issue now. I've added this line to the `build.rs` file:
```rust
    println!("cargo::rerun-if-changed=fonts/*.svg");
```
Thinking this would trigger a build whenever we changed an svg file, because previously if I only made a change on an svg file it wouldn't trigger a new build. And the line above sort of worked, as in it did started building whenever I've made a change to an svg file. However I've just noticed that now it is making a new `iced_fontello` build even when I make a change to the `example/src/main.rs` file which it shouldn't!

If I remove that line it stops doing it... However I think it is important to trigger a build whenever a change is made to an svg file!

Do you know any workaround for this? I've just learned about this `cargo::rerun-if-changed` a couple of days ago when I saw it on `iced_fontello`! 😄

alex-ds13 committed:
fix: remove wrong iced icon and add cargo reruns

- Removes the wrong iced icon "iced_logo".
- Add `cargo::rerun-if-changed` lines for each svg file, since it
  doesn't accept wildcards, using the directory still triggers a build
  on any change. So the only way to make it work correctly is to add a
  line for each svg file. If the user doesn't intend to change the svg
  files without changing the TOML file as well then these lines for each
  svg can be omitted.

alex-ds13 committed:
fix: remove wrong iced icon and add cargo reruns

- Removes the wrong iced icon "iced_logo".
- Add `cargo::rerun-if-changed` lines for each svg file, since it
  doesn't accept wildcards, using the directory still triggers a build
  on any change. So the only way to make it work correctly is to add a
  line for each svg file. If the user doesn't intend to change the svg
  files without changing the TOML file as well then these lines for each
  svg can be omitted.

alex-ds13 committed:
fix: stop generating different hash on rebuild

- Previously the custom-icon glyph uid was being randomly generated,
  this meant that when rebuilding the uid for the same custom-icon would
  always be different so the font hash would always change as well. Now
  we use an hash of the icon path as the uid so the icon uid is always
  the same and the font hash stays the same as well.

alex-ds13 committed:
fix: stop generating different hash on rebuild

- Previously the custom-icon glyph uid was being randomly generated,
  this meant that when rebuilding the uid for the same custom-icon would
  always be different so the font hash would always change as well. Now
  we use an hash of the icon path as the uid so the icon uid is always
  the same and the font hash stays the same as well.

alex-ds13 commented:
@hecrj I've finally got the time to refactor one of my projects to use `iced_fontello` with this PR and discovered that there was a bug where when using custom-icons it was always creating a different hash on rebuild. This is now fixed on the last commit I've just pushed. From using this on my own project this PR's version seems to be working quite well!

I'm posting this just in case you want to revisit the idea of merging this.

hecrj commented on src/svg_parser.rs:
```diff
@@ -0,0 +1,153 @@
+use std::{collections::BTreeMap, path::Path};
+
+use svg::{
+    node::element::tag::{self, Type},
+    parser::Event,
+};
+use svg_path_ops::pt::PathTransformer;
```
Any chance `usvg` may make this easier for us here?

hecrj replied:
I was wondering if `usvg` already simplified these paths for us.

alex-ds13 replied:
> I was wondering if `usvg` already simplified these paths for us.

So I've finally had some time to look into this.

The first time I looked at `usvg` I didn't see that it used the `tiny_skia_path` internally for the actual paths and paths operations. Knowing this, I'm able to use the paths from `usvg` and apply it's transforms. It has the added benefit that the function `path.abs_transform()` gets the absolute transformations for a path with all the transforms from it's parents included. Also `usvg` makes all paths absolute so turning the path into a string is simpler. *(I couldn't find a way to turn the final path into a string, except for the `Debug` impl for the `tiny_skia_path::Path`, so I've copied that to the `write_path` function)*

In the end we can also apply a translate and scale transform to the final path like we were doing before to make the glyph have the size that fontello expects and be on the correct position.

So I guess in the end this did simplify things quite a bit as the code seems simpler to follow and understand now.

Let me know what you think!

alex-ds13 committed:
refactor(custom-icons): use 'usvg' crate

- Use `usvg` crate to parse and transform the paths instead of the `svg`
  and `svg_path_ops` crates.
- Simplify the `svg_parser` using the `usvg` crate and its internal
  `tiny_skia_path` for path operations.

alex-ds13 committed:
refactor(custom-icons): use 'usvg' crate

- Use `usvg` crate to parse and transform the paths instead of the `svg`
  and `svg_path_ops` crates.
- Simplify the `svg_parser` using the `usvg` crate and its internal
  `tiny_skia_path` for path operations.
